src/tests/ke/suite/thread.c

```diff
@@ -0,0 +1,250 @@
+#include <xboxkrnl/xboxkrnl.h>
+#include <processthreadsapi.h>
+#include <stdio.h>
+
+#include "util/output.h"
+#include "util/misc.h"
+#include "util/thread.h"
+#include "assertions/defines.h"
+
+typedef ULONG NTAPI (*threadFuncAPI) (PKTHREAD);
+
+typedef struct _thread_test {
+    // Previous result check
+    BOOL addThreadCounter;        // 0 or 1
+    DWORD expected_resultWait;    // WAIT_TIMEOUT or WAIT_OBJECT_0
+    // Next result task
+    threadFuncAPI nextFuncAPI;    // KeResumeThread or KeSuspendThread
+    // Don't need individual check result value, it is always STATUS_SUCCESS or
+    // STATUS_INVALID_HANDLE for wrong handle input
+    ULONG expected_suspendCount;  // Return suspend count after call
+    ULONG expected_threadCounter; // counter + addThreadCounter dynamic
+    ULONG return_threadCounter;
+    DWORD return_resultWait;
+    ULONG return_suspendCount;
+} thread_test;
+
+typedef struct {
+    BOOL run;
+    ULONG counter;
+    HANDLE hEventSecondary;
+    HANDLE hEventPrimary;
+} thread_counter_s;
+
+static DWORD NTAPI KeResumeSuspendThread_secondary(void* arg)
+{
+    thread_counter_s* thread_data = (thread_counter_s*)arg;
+    do {
+        THREAD_DUO_EVENTS_WAIT_SECONDARY(thread_data->hEventSecondary);
+        thread_data->counter++;
+
+        // Let other thread to run
+        THREAD_DUO_EVENTS_NOTIFY(thread_data->hEventSecondary, thread_data->hEventPrimary);
+    } while(thread_data->run);
+    return 0;
+}
+
+static BOOL KeResumeSuspendThreadInline(const char* test_name, BOOL suspend, thread_test* thread_tests, unsigned total)
+{
+    ASSERT_HEADER;
+
+    THREAD_DUO_EVENTS_CREATE(hEventMain, hEventThread);
+
+    thread_counter_s thread_data = {
+        .run = TRUE,
+        .counter = 0,
+        .hEventSecondary = hEventThread,
+        .hEventPrimary = hEventMain
+    };
+
+    ULONG old_suspend_count, counter = 0;
+    NTSTATUS result;
+    DWORD result_wait;
+
+
+#if 0 // TODO: Could not catch invalid argument at this time.
+    if (suspend) {
+        BOOL catch = -1;
+        __try {
+            (void)KeSuspendThread(NULL);
```

Comment:
These tests are wrong. `KeSuspendThread` only raises an exception when the suspend count of the thread exceeds the maximum limit, and expects the thread object to always be valid (because it's supposed to have already been validated by `NtSuspendThread`). Passing an invalid object like this will cause a page fault in the kernel and there's no point in testing that.

Reply:
I assumed you missed few lines above?
`#if 0 // TODO: Could not catch invalid argument at this time.`

These codes aren't enabled by default due to unfinished work. The purpose of these tests are making sure they do break as seen in `GEN_CHECK(catch, 1, "catch");` line as Ke functions here are meant to check they are not safe-checking the argument inputs. Only the Nt functions from this pull request does have safe-checking the argument.

The whole purpose of this kernel test suite is to "brute force" validate that they work by design as intended by the official Xbox kernel.

> `KeSuspendThread` only raises an exception when the suspend count of the thread exceeds the maximum limit.

I did not know nor checked what was the real limit as I did not get to that point yet.

Reply:
> I assumed you missed few lines above?

I didn't, but it will become an issue when you eventually re-enable this code, so I thought to report it anyway.

> The purpose of these tests are making sure they do break as seen in GEN_CHECK(catch, 1, "catch"); line as Ke functions here are meant to check they are not safe-checking the argument inputs

I disagree. As a general rule of thumb, the test suite should only test conditions that the functions expect and can handle/recover from. Passing a nullptr here will cause a crash in the kernel, and anything that happens after that is undefined behavior. So, test results from such situations are invalid by definition in my opinion.